Enhancement to Resilience Metrics

yujiex opened the pull request:
Pull request overview
---------------------
With the increasing frequency and severity of extreme weather events (e.g., heat waves), it is crucial to ensure urban buildings and infrastructure are resilient to provide critical services to preserve human life and properties during natural disasters. Building resilience could become an additional value proposition for technologies and systems if it can be reliably quantified, valued, and trusted by stakeholders. Resilience metrics or an assessment of potential vulnerability, likelihood, and consequence for each risk would help to prioritize further consideration of these risk factors of building resilience. 

EnergyPlus currently calculates a set of thermal, visual, and indoor air quality (IAQ) resilience metrics as optional reportable variables. The resilience reports are presented as three tabular summary reports (Thermal Resilience Summary, Indoor Air Quality Resilience Summary, and Visual Resilience Summary) and can be requested as part of the tabular reports. The thermal resilience report includes Heat Index, Humidex, and Standard Effective Temperature (SET) hours as three separate metrics. To support the need for new resilience metrics in building simulation and to provide a more convenient path in adopting existing metrics, we propose to (1) calculate and report Hours of Safety, Setpoint Degree Hours Unmet, Discomfort Weighted Hours as three additional building thermal resilience metrics following EPA and ASHRAE Standard 55–2020 guidance, and (2) improve current resilience metrics reporting by allowing users to specify the period of calculating and reporting the metrics which is not necessarily the same as the simulation period.

## Major modifications include
- Added the following new report tables for run period (annual) thermal resilience tabular reports
  - Hours of Safety for Cold/Heat Events
  - Unmet Degree-Hours
  - Discomfort-weighted exceedance occupant/occupied hours
- Added thermal, CO2, visual resilience tabular report for reporting periods
- Two new numeric fields in the idd object *People*, holding temperature thresholds for hours of safety reporting
- A new input object *Output:Table:SummaryReports:ReportPeriod* specifying the start and end time of the reporting period
- Adding an "Occupied hours" column to all reporting tables. Previously, there is only an "OccupantHours" (accumulated hours for the sum of all occupants in a space). The new column "Occupied hours" will accumulate measurments for total occupied hours without multiplying by the number of occupants. The two columns will become three columns as follows
  - Degree-hours summed for all hours. Currently, the eplusout.csv reports SET for all timestamps regardless of the number of occupants in the zone. The eplusout.htm "SET > 30°C Hours (°C)" column, for example, will only consider hours with `Number of People x Number of People Schedule >= 1`. This is a bug to be fixed.
  - Degree-hours summed for all occupied hours (num of occ * schedule >0)
  - Occupant-degree-hours summed for all occupied hours weighted by the actual number of occupants (num of occ * schedule >0)
- For hours of safety table, only keep the "Save Temperature Exceedance Hours" column. Do not consider OccupantHours.
- Add unit conversion for the thermal resilience tabular report. Previously, whether the "Unit Conversion" field in OutputControl:Table:Style object only change the header of the resilience reporting table, but not the contents.
- Changed header wording (degree-hour) and units (°C to °C·hr) for SET degree-hour tables
- Added a warning for thermal resilience tables when one zone has multiple people objects and they don't agree on metrics like PMV, PierceSET, or Cold/Heat Stress Temperature Threshold.

## Example output 

The links to the resilience reporting period tables will show up in the table of contents
<img width="776" alt="image" src="https://user-images.githubusercontent.com/13421161/172963706-a4474bc9-6e09-4e6b-bfd8-857244f44a74.png">

The reporting period time range will be shown at the beginning of each reporting period for each type of resilience report, 
<img width="1662" alt="image" src="https://user-images.githubusercontent.com/13421161/172963806-4cfdc403-6978-4fb0-9d73-152e8d66512a.png">

## Regression diffs
- 513 files have BIG tablediffs: the difference are in the "Defaulted Fields" and "Fields with Defaults". This is due to the addition of two input fields: `Cold Stress Temperature Threshold` and `Heat Stress Temperature Threshold` in the `People` object.
- 723 files have AUDIT differences: there are increases in numSubTable, numColumnTag, and numTableEntry. This is due to the addition of new resilience tables in the annual thermal resilience summary, new "*OccupiedHour" columns in existing resilience reporting tables.
- 2 files have ERR diffs: this is due to an added warning message when multiple people objects are defined for the same zone.


## Examine heat index and hours of safety during a power outage with vs without natural ventilation.

Two idf files are used in the comparison: `1ZoneEvapCooler_ReportPeriodPowerOutage.idf` and `1ZoneEvapCooler_ReportPeriodPowerOutage_naturalVent.idf`. The latter accounts for natural ventilation using a `ZoneVentilation:DesignFlowRate` object defined as follows, following the 5ZoneNightVent2 example, except that the NightVentSched value is 1 throughout the reporting period: 7/25 8:00 -- 7/27 18:00.

```
  ZoneVentilation:DesignFlowRate,
    Main Zone Ventl,         !- Name
    Main Zone,               !- Zone or ZoneList Name
    NightVentSched,          !- Schedule Name
    flow/zone,               !- Design Flow Rate Calculation Method
    0.05295,                 !- Design Flow Rate {m3/s}
    ,                        !- Flow Rate per Zone Floor Area {m3/s-m2}
    ,                        !- Flow Rate per Person {m3/s-person}
    ,                        !- Air Changes per Hour {1/hr}
    Intake,                  !- Ventilation Type
    67.,                     !- Fan Pressure Rise {Pa}
    0.7,                     !- Fan Total Efficiency
    1,                       !- Constant Term Coefficient
    0,                       !- Temperature Term Coefficient
    0,                       !- Velocity Term Coefficient
    0,                       !- Velocity Squared Term Coefficient
    18,                      !- Minimum Indoor Temperature {C}
    ,                        !- Minimum Indoor Temperature Schedule Name
    100,                     !- Maximum Indoor Temperature {C}
    ,                        !- Maximum Indoor Temperature Schedule Name
    2.0,                     !- Delta Temperature {deltaC}
    ,                        !- Delta Temperature Schedule Name
    -100,                    !- Minimum Outdoor Temperature {C}
    ,                        !- Minimum Outdoor Temperature Schedule Name
    100,                     !- Maximum Outdoor Temperature {C}
    ,                        !- Maximum Outdoor Temperature Schedule Name
    40;                      !- Maximum Wind Speed {m/s}
```

The following figure shows the outdoor dry-bulb temperature, the indoor mean air temperature, and the ventilation rate. We can observe a slightly lower indoor mean temperature with natural ventilation. There is a two-hour reduction in the caution-range heat index with natural ventilation, a 0.16-hour increase in hours of safety, and some reduction in total safety temperature exceedance hours.

![cmp_with_without_vent](https://user-images.githubusercontent.com/13421161/180899121-1183e619-722f-4df6-a929-31f1c213702c.png)

The following compares the Heat Index and Hours of Safety
<img width="929" alt="image" src="https://user-images.githubusercontent.com/13421161/180901469-3883446b-89a2-444b-b64e-4eddafc6c0c4.png">
<img width="942" alt="image" src="https://user-images.githubusercontent.com/13421161/180901528-ba4225ca-4a81-4732-896d-2eaedc0ff0c7.png">


**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [x] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [x] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

yujiex commented:
Hello @Myoldmopar @mjwitte @JasonGlazer, would you be available to review this PR and provide some feedback? Thanks :)

mjwitte commented on design/FY2022/enhacement_to_resilience_metric.md:
```diff
@@ -0,0 +1,191 @@
+
+We will add a new object *Output:Table:SummaryReports:ReportPeriod* to add supplementary information to define the reporting period of certain summary tables.
+
+    Output:Table:SummaryReports:ReportPeriod,
```
I would suggest `OutputControl:Table:ReportPeriod`, similar to `Output:Table:Style`.

yujiex replied:
The intent of the "SummaryReports" indicates the reporting period is only for the tabular reports, but I agree we should follow the "Output:Table:Style". I will change "Output:Table:SummaryReports:ReportPeriod" to "OutputControl:Table:ReportPeriod" in the coming push.

mjwitte commented on design/FY2022/enhacement_to_resilience_metric.md:
```diff
@@ -0,0 +1,191 @@
+
+    Output:Table:SummaryReports:ReportPeriod,
+        \memo This object allows the user to add a reporting period to a certain tabular report when the reporting period is not the whole simulation period. When defined, the declared summary report is based on aggregations from the start date to the end date defined. 
+      A1, \field Report Name,
```
If I'm following the intent correctly, a user may want to specify more than one reporting period. e.g. a summer power outage and a winter power outage. In that case, then field A1 should simply be Name (of the object), and Report Name would be A2. Another option would be to use a reporting schedule in this object instead of dates and times.  Either way, it will be important for the table outputs to include a note about the dates and number of hours the report was active.

yujiex replied:
Good point. We'll have some internal discussion about how to go about enabling multiple reporting periods and update the feature design doc on this topic sometime next week. 

In the tabular report, under each table, a line of text, "Reporting period: xxx-xx to xxx-xx", is added to show what reporting period this table has. This line is a placeholder for now. It will be updated once we figure out the multiple report period design.

yujiex replied:
Hi @mjwitte, the "Name" field is added to the idd object to allow for multiple reporting periods.

JasonGlazer replied:
I am concerned that OutputControl:Table:ReportPeriod is going to imply to people not assessing resilience at all that these are the report periods for all of the output tables (ABUPS, etc). I presume that it is not your intention to modify all output table data collection periods based on this input, correct? If so, I think the name needs to be even more specific, how about: "OutputControl:Table:ResilienceReportPeriod"

yujiex replied:
Hi @JasonGlazer, this ReportPeriod object is not specific to the resilience metric. Although currently we only allow for three keys (or maybe with the fourth one of AllResilienceMetric), we want it to be more generic and can be applied to other tables like energy reporting. For example, with this ReportPeriod feature, we can report energy usage by season.

mjwitte commented on design/FY2022/enhacement_to_resilience_metric.md:
```diff
@@ -0,0 +1,191 @@
+
+N/A
+
+## Proposed Report Variables: ##
```
Would any of these new measure be useful as a time series output variable?

yujiex replied:
Maybe some variables like "unmet degree hours" and "discomfort weighted exceedance hours" could potentially be useful as a time series output. Hours of safety may be less meaningful as a standalone time series output, as it measures _the time_ a zone first crosses the dangerous temperature threshold. However, the implementation timeline is a bit tight. We're trying to start simple. Maybe the addition of other time series output variables could be made in the next feature update?

mjwitte replied:
@yujiex Of course, it's fine to wait to add output variables later. And since you are on a tight schedule, the zone daylighting question can wait also.

yujiex replied:
Thanks, @mjwitte. We'll leave this time series variable reporting and the daylighting part to future development.

yujiex replied:
Hi @mjwitte, we were considering adding some variables to the advanced report variable set for debugging purposes.

mjwitte reviewed (commented):
This looks good in general. A few comments below. Also, when I was working on the daylighting code for Space, I kept the [`ZoneDaylightCalc` struct](https://github.com/NREL/EnergyPlus/blob/b448a8250a1db9995f7f7f91cfe686e85f956a3f/src/EnergyPlus/DataDaylighting.hh#L214-L218) solely to support the visual resilience report. Please review how that report is calculated and we can talk about whether it would make more sense to be calculated at the Space or Enclosure level. Maybe it's fine as-is, but it's worth reviewing.

EnergyArchmage commented on design/FY2022/enhacement_to_resilience_metric.md:
```diff
@@ -0,0 +1,200 @@
+          \minimum 1
+          \maximum 31
+          \type integer
+      N3 , \field Begin Hour of Day
```
I would think may also want begin year and end year to be specified.

yujiex replied:
Thanks. Just added the "begin year" and "end year" fields in the ReportPeriod object.

yujiex committed:
Add name to ReportPeriod, add advanced reporting variables

yujiex committed:
Typo

yujiex committed:
Corrected the term for the advanced report variable

JasonGlazer commented on design/FY2022/enhacement_to_resilience_metric.md:
```diff
@@ -0,0 +1,211 @@
+          \key ThermalResilienceSummary
+          \key CO2ResilienceSummary
+          \key VisualResilienceSummary
+          \key …
```
Perhaps add an option for "AllResilienceSummaries"

yujiex replied:
Thanks, just added this key :)

JasonGlazer commented on design/FY2022/enhacement_to_resilience_metric.md:
```diff
@@ -0,0 +1,211 @@
+
+The default "*Cold Stress Temperature Thresholds*": 60°F (16°C) for the healthy population [1]. Other values can be used: 64°F (18°C) for the elderly [1], and 71°F (22°C) for nursing home residents [2]. The default "*Hot Stress Temperature Threshold*" is selected as 86°F (30°C) [2]. Users can modify these default thresholds with the "*Cold Stress Temperature Threshold*" and "*Heat Stress Temperat ...
+
+We will report both the zone and building-level Hours of Safety. For the zone level, we will report the longest duration in a risk period a zone can maintain comfort and safety before reaching unsafe indoor temperature thresholds. For the building level, we’ll provide four summary statistics across all zones in the building: min, max, average, and sum.
```
Please add to the documentation that during events with power outages that models should add strategies that are typically  used by people in these circumstances such as increased airflow from window usage during hot events and the use of non-electric heating (fireplaces, stoves, etc.) during cold events. A model that does not take these considerations into account would be misleading. Especially for the hot events, EnergyPlus will show a sealed room get very hot.

yujiex replied:
Thanks, the "Approach" section is updated about documentation updates suggesting users consider human adaptation strategies during hot and cold events in their models.

JasonGlazer commented on design/FY2022/enhacement_to_resilience_metric.md:
```diff
@@ -0,0 +1,211 @@
+Resilience metrics are more often evaluated during a certain period when a building is at risk (e.g., during the power outage event or heatwave event), and the period is not necessarily the same as the whole simulation period. We propose to improve current resilience metrics tabular reporting (in the html output) by allowing users to specify the period of calculating and reporting the metrics. 
+
+We propose to add an *Output:Table:ReportPeriod* that takes an input of a Summary Report name (e.g. *ThermalResilienceSummary*). The reporting period can be defined with a start-date field and an end-date field as supplementary inputs to the *ThermalResilienceSummary* report.
+
```
I would recommend adding a warning message if electricity is consumed at the facility during the timespan defined. This is intended to model power outages.

yujiex replied:
Indeed, I will add this check of non-zero electricity usage during a power outage.

yujiex committed:
Add another key for AllResilienceSummaries in ReportPeriod key

yujiex committed:
Add non-zero-elec check warning, doc about adaptation in power outage

- warning: Jason recommended adding a warning message if electricity is consumed
  at the facility during the timespan defined. This is intended to model power
  outages
- doc: Per Jason's suggestion: add to the documentation that during events with
  power outages that models should add strategies that are typically used by
  people in these circumstances such as increased airflow from window usage
  during hot events and the use of non-electric heating (fireplaces, stoves,
  etc.) during cold events. A model that does not take these considerations into
  account would be misleading. Especially for the hot events, EnergyPlus will
  show a sealed room get very hot.

nrel-bot-2c commented:
@yujiex @myoldmopar it has been 28 days since this pull request was last updated.

nrel-bot-3 commented:
@yujiex it has been 16 days since this pull request was last updated.

nrel-bot-3 commented:
@yujiex @myoldmopar it has been 28 days since this pull request was last updated.

yujiex committed:
Add heat and cold stress temperature thresh to idd people

yujiex committed:
Read in people input of cold and heat temp thresh

yujiex committed:
Add pre-defined table structure for hours of safety

yujiex committed:
Add hour of safety table

yujiex committed:
Add unmet degree hour table

yujiex committed:
Add Discomfort-weighted Exceedance Hours table

yujiex committed:
Change hours of safety calc to people loop

yujiex committed:
Add Output:Table:ReportPeriod in idd

yujiex committed:
Add notes to report name fields

yujiex committed:
Add code reading inputs of Output:Table:ReportPeriod

yujiex committed:
Find which reporting period it is in

yujiex committed:
Add heat index table for reporting period

mjwitte commented:
@yujiex I see some comments were addressed in November. Is this ready for a final review? Please merge in develop to bring this branch up to date.

nrel-bot-3 commented:
@yujiex @myoldmopar it has been 28 days since this pull request was last updated.

yujiex committed:
Mod HI table for reporting period

- fix index error in filling the table content
- add table of contents entries
- move the reporting period tables to the end
- change reporting period formatting to differ when year is present or not,
  so that dates like 0/03/05 will show only month and day

yujiex committed:
Add a test file for reporting period test

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Fix findReportPeriodIdx signature, no need to pass in dates

yujiex committed:
Add cold heat stress temp thresh in unit test object

yujiex committed:
clang-format

yujiex committed:
Modify num of people to show difference in occupant and occupied hr

yujiex committed:
separate out a function to write table for reporting period

yujiex committed:
Add occupant and occupied hour table for heat index hours

yujiex committed:
Fix integration tests due to added idd fields

yujiex committed:
Add new test idf to CMakeLists.txt

yujiex committed:
clang-format

yujiex committed:
Added deallocation for heat index hours report period tables

yujiex committed:
Fix integration test issue due to added idd fields

yujiex committed:
Add humidex hours tables

yujiex committed:
clang-format

yujiex committed:
Fix html SET hours always reporting occupant hours

- rename ZoneOccPierceSet to ZonePierceSet as there is a column calculating the
  Pierce SET hours for the whole period.
- remove the NumOcc > 0 check in the table entry calculation and the
  ZoneNumOcc(ZoneNum) check in getting value to ZonePierceSet.

yujiex committed:
Fix more integration test issues

yujiex committed:
Add occupied one col in SETHour table, add SETHour for report period

yujiex committed:
clang-format

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Use standardized unit of measure

yujiex committed:
Add occupied hours column in unmet deg hour annual table

yujiex committed:
Add unmet degree hour reporting period table

yujiex committed:
Rename degree-hour related table title and col header

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Give warning for multiple people objects for one zone

yujiex committed:
Add occupied col in hours of safety

also remove the "/ max(1, NumOcc)", wrong approach to account for multiple
people objects

yujiex committed:
Add report period table for hours of safety

yujiex committed:
Discomfort-weighted exc hr separate into occupant and occupied

yujiex committed:
Add discomfort-weighted hr reporting period table

yujiex committed:
remove unused init

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Add another people object to zone 1 to trigger warning

yujiex committed:
Update design document

yujiex committed:
clang-format

yujiex committed:
Add unit conversion to heating cooling SET deg-hr

yujiex committed:
Add unit conv, hasPierceSET, WriteThmRes check in report period tbls

- add unit conversion in both the run period tables of SET and Unmet
degree-hours table
- only write heating cooling SET degree-hour table when hasPierceSET is true
- put WriteThermalResilienceTables() call under the check of
displayThermalResilienceSummary

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
clang-format

yujiex committed:
Remove real-to-int casting of number of occupant, NumOcc

yujiex committed:
Add occupied hour table for sim period HI and Humidex

yujiex committed:
Unit test for reading new idd fields and object

yujiex committed:
Change ZoneNumOcc to float to prevent casting

yujiex committed:
Fix hr of safety bug, fix set longest hr acc bug

- hr of safety accumulator was in the wrong branch, fixed
- SET need to only update longest hour when it is occupied

yujiex committed:
Add more tests to run period resilience metric tbls

- add tests for new tables
- add tests for occupied and occupant hrs, considering occupancy in (0, 1)

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Change col header longest set degree-hr and start time for occu

Longest SET > 30C hours only count occupied period, so is start time

yujiex committed:
Fix HOS rep period acc in wrong branch, acc SET longestHr in occu

- move the temperature exceeding hours accumulator into the "danger" branch
- accumulate longest SET > 30 or <= 12.1 only for occupied hours

yujiex committed:
Unit test for reporting period resilience table

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Add test for warning: multiple people defined for one zone

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Fix unitConv mult on hr column, expose tableBody for testing

- unit conversion in SET Degree-Hours tables used to multiply the *Duration [hr]
  as well. This is wrong. Change multipliers to 1 for this column.
- add rowHead and tableBody in the Write...ReportingPeriod() function argument
  list, so that during testing, the entries can be retrieved easily
- added a function to RetrieveEntryFromTableBody to get entries from tableBody

yujiex committed:
Added unit conversion tests for run period and reporting period tbls

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Update comments in two resilience table tests

yujiex committed:
clang-format

yujiex committed:
Fix tableBody size issue.

yujiex committed:
Add CO2 and Visual report period tbl, add occupiedHr table

yujiex committed:
Initial commit of reportperiod idd documentation

yujiex committed:
Change idd memo of ReportPeriod

yujiex committed:
Fix table of contents for report name == AllResilienceSummaries

yujiex committed:
Add documentation to Cold/Heat Stress Temp Thresh idd in People

yujiex committed:
Fixed typo in comments

yujiex committed:
Added another testfile for "AllResilienceSummaries" option

yujiex committed:
Added tests for CO2 and Visual reporting period tables

yujiex committed:
Fix unit test table acc allocation in TestResilienceMetricReport

yujiex committed:
Change the order of occupant and occupied table for HI and humidex

yujiex committed:
Add documentation

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
clang-format

yujiex commented:
Hi @JasonGlazer @EnergyArchmage @mjwitte the PR is ready for review. Please see if the implementation and documentation updates look good and if anything needs to be modified. Thanks :)

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Add cold and heat stress temp thresh in people in a unit test

yujiex committed:
clang-format

JasonGlazer commented:
@yujiex have you looked at all the differences to AUD, Table, ERR, and IDF and are they what you expected?

yujiex committed:
Group report period by report type, resolve overlapping report period

Previously findReportPeriodIdx only returns an index of the reporting period,
this requires the reporting period to not overlap. Now change it to an array of
T/F to indicate whether a timestamp belongs to a reporting period. Re-indexing
reporting period within each report type, this fixes the indexing issue.

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Add unit test for visual reporting period tables

yujiex committed:
Unnecessary casting removed

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex commented:
Just fixed the reporting indexing issue and overlapping reporting period issue.

yujiex commented:
Following summarizes the regression diffs, I also added them in the PR overview @JasonGlazer 
* 513 files have BIG tablediffs: the difference are in the "Defaulted Fields" and "Fields with Defaults". This is due to the addition of two input fields: `Cold Stress Temperature Threshold` and `Heat Stress Temperature Threshold` in the `People` object.
* 723 files have AUDIT differences: there are increases in numSubTable, numColumnTag, and numTableEntry. This is due to the addition of new resilience tables in the annual thermal resilience summary, new "*OccupiedHour" columns in existing resilience reporting tables.
* 2 files have ERR diffs: this is due to an added warning message when multiple people objects are defined for the same zone.

JasonGlazer commented:
@yujiex I started to do a review. Concerning the new warning, having more than one People object is not really the problem, right? It is multiple People objects that have different stress temperature setpoints, right? If that is correct, could you change when the warning is generated to only when multiple People objects are used in a zone when they have different stress temperature setpoints?

The Clang format issues should probably be resolved prior to a deep review.

yujiex commented:
> @yujiex I started to do a review. Concerning the new warning, having more than one People object is not really the problem, right? It is multiple People objects that have different stress temperature setpoints, right? If that is correct, could you change when the warning is generated to only when multiple People objects are used in a zone when they have different stress temperature setpoints?

Right. The point is to prevent zone-level aggregation of people with different thresholds. I'll change the warning to only when they have different PMV or different SET or different Heat or cold stress temperature threshold. I'll also fix the clang format.

yujiex committed:
Add warning msg for multiple people different thresh in one zone

yujiex committed:
Update test with warning changes in multiple people in 1 zone

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
clang-format

yujiex commented:
Hi @JasonGlazer, I've changed the warning messages to be produced only when different values of PierceSET, PMV, or Cold/Heat Stress Temperature Threshold are encountered

yujiex committed:
clang-format

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Change std::optional to special value, remove vec alloc

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
clang-format

yujiex committed:
clang-format

yujiex committed:
Update documentation and comments

- warnings are produced when multiple people are having different SET, PMV etc.
- caution when reporting over power outage periods.

yujiex committed:
Update md with change in warnings

yujiex committed:
Add total electricity consumption for each reporting period

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex commented:
Hi @JasonGlazer I just added in the total electricity use for the reporting period next to the time range as follows
<img width="1662" alt="image" src="https://user-images.githubusercontent.com/13421161/173440407-e3972d8e-1de4-45e4-a9a2-738019fa70c7.png">
Please see if there are other issues to address. Thanks

yujiex committed:
clang-format

JasonGlazer commented:
> Hi @JasonGlazer I just added in the total electricity use for the reporting period next to the time range as follows <img alt="image" width="1662" src="https://user-images.githubusercontent.com/13421161/173440407-e3972d8e-1de4-45e4-a9a2-738019fa70c7.png"> Please see if there are other issues to address. Thanks

@yujiex I really like this idea. Is the reporting period dates or the electricity showing up in the SQLite or Output:JSON version of the output? If not, you might want to make just a small table just so the output is available in multiple formats.

yujiex commented:
> > Hi @JasonGlazer I just added in the total electricity use for the reporting period next to the time range as follows <img alt="image" width="1662" src="https://user-images.githubusercontent.com/13421161/173440407-e3972d8e-1de4-45e4-a9a2-738019fa70c7.png"> Please see if there are other issues to address. Thanks
> 
> @yujiex I really like this idea. Is the reporting period dates or the electricity showing up in the SQLite or Output:JSON version of the output? If not, you might want to make just a small table just so the output is available in multiple formats.

Hi @JasonGlazer, I will double-check and make sure the SQLite or JSON output also has it. Currently, I'm trying to hunt down the two integration errors with some access-out-of-bound issues. I'll update you when the reporting-time total electricity SQLite or JSON output is added. Thanks :)

yujiex committed:
Resolve array out-of-bound error

yujiex committed:
Make the temperature thresh for testing error a comment in test idf

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Fixing the unit test with Cold/Heat temp thresh from merging

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Fix test idfs adding cold/heat stress temp thresh

yujiex committed:
Add sqlite reporting for reporting period resilience report

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Missing comma

Myoldmopar commented:
This needs transition rule descriptions and transition code, right?

yujiex committed:
Add a period summary table to html and sqlite output

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Fix unit test for the idd change

yujiex commented:
> This needs transition rule descriptions and transition code, right?

@Myoldmopar Yes, indeed. The idd change will need some transition code. I'll look into them.

Myoldmopar commented:
> @Myoldmopar Yes, indeed. The idd change will need some transition code. I'll look into them.

The ideal scenario (for me) is when the individual developers do the transition rules and code themselves, but I concede that the development team is not as up to speed with Fortran as it once was.  So if you can write out the rules in the `src/Transition/InputRulesFiles` file, I can do the Fortran implementation of those rules.

yujiex commented:
> > @Myoldmopar Yes, indeed. The idd change will need some transition code. I'll look into them.
> 
> The ideal scenario (for me) is when the individual developers do the transition rules and code themselves, but I concede that the development team is not as up to speed with Fortran as it once was. So if you can write out the rules in the `src/Transition/InputRulesFiles` file, I can do the Fortran implementation of those rules.

Yes, I'll do it now. There's a lot of output change as well, I'll also add those in the OutputRulesFiles, but I'll push the input ones first. Thanks

yujiex committed:
input transition rule change

yujiex commented:
Hi @Myoldmopar, I just pushed the input transition rules. Please see if it is clear. Thanks.

yujiex committed:
input transition rule following older style

yujiex committed:
output tabular report change updated in output transition doc

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

JasonGlazer committed:
Merge branch 'develop' into enhanceResilienceMetrics

JasonGlazer commented:
@yujiex sorry about the delay. I took a close look at the code and ran a few test files. I have the following observations:
 - Why is the “Reporting Period Time and Consumption” table in Demand End Use Components Summary report? I think it may show up there by accident.
 - The IDD change uses the term “threshold” but the input files all just say “thresh” they should be changed to “threshold” to match the IDD.
 - We need an example file that does have a power outage at the same time as the reporting period. Have you figured out how to do a power outage? You might be able to use ElectricLoadCenter:Transformer Availability Schedule but I have never tried this. Whatever you figure out, I think we need documentation on how to simulate a power outage when it is mentioned for the People object and for the section of the IOref that describes Output:Table:ReportPeriod.
 - I’m still uncomfortable reporting thermal resilience properties during a hot weather power outages that don’t include natural ventilation from opening windows. In my experience, EnergyPlus sometimes does weird things when air conditioning is not running and spaces can get to very hot temperatures that are unrealistic. Have you tested how the temperatures climb in a room without operable windows and with operable windows and natural ventilation? Do either seem reasonable? While this question may be outside of your scope, it concerns me that EnergyPlus will be used to show buildings not being resilient that really might be fine since occupants will open windows.
 - Are you being careful to make sure these calculations are only performed when requested? It seems like the simulation is a little slower now than before.
 - Seems like isReportPeriodBeginning and findReportPeriodIdx should be in general.cc instead of HeatBalanceSurfaceManager.cc
 - The SI units at the top of the HeatIndexOccupiedHours (any other places) and not translated into IP when InchPound output is specified.
 - You seemed to do a good job with adding unit tests but I did not see any specific tests for some of the new functions that you added that seem like utility functions: BetweenDateHoursLeftInclusive, BetweenDateHours, isReportPeriodBeginning, findReportPeriodIdx, AddTOCReportPeriod, RetrieveEntryFromTableBody, CopyReportPeriodObject

Overall, I am good with code and documentation changes besides the comments above.

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Merge remote-tracking branch 'origin/enhanceResilienceMetrics' into enhanceResilienceMetrics

yujiex committed:
Heat/Cool... Thresh changed to "Threshold"

yujiex committed:
Add Reporting Period Time and Consumption to Rep Period Summary

yujiex committed:
Move isReportPeriodBeginning and findReportPeriodIdx to general.cc

yujiex committed:
fix header IP unit setting for report period tbls

yujiex committed:
Modify IP unit settings in column headers for annual table

have to change all the annual thermal resilience tables to not use predefined
table, as interleaving will make them show up in two different places.

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Fix columnWidth not init bug

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
report period summary & toc entry only when report period exist

- add checks to produce report period TOC entries and report period summary only
  when the number of reporting period > 0

yujiex committed:
Fix tableBody dimension for unmet deg hour

yujiex committed:
Add heat/cold stress temperature threshold for testfile idfs

yujiex committed:
Short-circuiting limit calculation to only when rep period is defined

yujiex committed:
change function name to match others not using predefined tbl

Myoldmopar commented:
CI is trying to play catch up today, but Mac already showed that there is a table issue.  I think the error message on the CI results page shows the issue pretty clearly.  Let me know if you need more help getting this clean so it can be ready to merge soon.

yujiex commented:
> CI is trying to play catch up today, but Mac already showed that there is a table issue. I think the error message on the CI results page shows the issue pretty clearly. Let me know if you need more help getting this clean so it can be ready to merge soon.

Yes, I'm fixing them now. The message is pretty clear to me and I can reproduce them on my local machine

yujiex committed:
Fix json html not align

yujiex committed:
clang-format

yujiex commented:
~> CI is trying to play catch up today, but Mac already showed that there is a table issue. I think the error message on the CI results page shows the issue pretty clearly. Let me know if you need more help getting this clean so it can be ready to merge soon.~

~Hi @Myoldmopar we're still investigating the differences in the test examples new.idf using the implemented multiplier vs orig.idf using the EMSActuator. The difference seems to come from the wall inside surface heat gain.~

Commented on the wrong PR

yujiex committed:
unit test BetweenDateHoursLeftInclusive

yujiex committed:
remove unused function

yujiex committed:
Add a false case for the BetweenDateHoursLeftInclusive test

yujiex committed:
Add test for isReportPeriodBeginning

yujiex committed:
Add test for findReportPeriodIdx

yujiex committed:
Add test for RetrieveEntryFromTableBody

Myoldmopar commented:
Looks like we're down to just regressions, and those are just table/audit diffs.  I assume this is very close.  Linux CI is still trying to catch up I think, so there are still some results to come, including debug runs.  @yujiex are you planning more commits here or will this one be ready for final review/merge if CI is happy?

yujiex commented:
> Looks like we're down to just regressions, and those are just table/audit diffs. I assume this is very close. Linux CI is still trying to catch up I think, so there are still some results to come, including debug runs. @yujiex are you planning more commits here or will this one be ready for final review/merge if CI is happy?

@Myoldmopar I'll double-check the documentation and test cases again and will probably be ready today or tomorrow.

yujiex committed:
Add test for CopyReportPeriodObject, fix reportName not copied

Myoldmopar commented:
Thanks, when you make any final changes, note that there is a Clang Format issue in the unit test code, so include a reformat for that stuff.

yujiex committed:
Correct the summary rows in SET degree hr table

yujiex committed:
Add unit tests for nonPreDef resilience table functions

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
resilience table with power outage

yujiex committed:
clang-format

yujiex committed:
Revert "clang-format"

This reverts commit 8c5b3ebd8fc9785c2fd62047c9aa65b5bbf754ad.

yujiex committed:
Revert "resilience table with power outage"

This reverts commit 2b13b13c4ed7c38f365d0b43f31bb0e8bde62c1c.

yujiex committed:
Testfile with reporting period and power outage during the period

yujiex committed:
Update comments in the test idf with power outage

yujiex committed:
clang-format

yujiex committed:
Add testfile to add_simulation_test in CMakeLists

yujiex committed:
fix indexing

yujiex committed:
Add occupant and change reporting period in test file

yujiex committed:
Resilience Metric testfile with poweroutage and natural ventilation

yujiex committed:
Add window, change back heat stress temperature threshold

yujiex committed:
Make ventilation always on, replace constant schedule with value

yujiex committed:
Add test idf for with ventilation

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex commented:
Hi @JasonGlazer,

I addressed the comments as follows 
- Put "Reporting Period Time and Consumption" in a separate section called "reporting period summary"
- Changed "thresh" to "threshold" in people objects in the test idfs, unit tests files etc.
- Added an example file with a power outage: "1ZoneEvapCooler_ReportPeriodPowerOutage.idf"
- Added some checks to make sure reporting period index checks only happen when the object is defined
- isReportPeriodBeginning and findReportPeriodIdx are moved to general.cc
- Change the column header to the corresponding IP or SI units for Heat Index Hours/Occupant Hours/Occupied Hours, Heating/Cooling SET Degree Hours.
- Added unit tests for BetweenDateHoursLeftInclusive, isReportPeriodBeginning, findReportPeriodIdx, RetrieveEntryFromTableBody, CopyReportPeriodObject, WriteResilienceBinsTableNonPreDef, WriteSETHoursTableNonPreDef, WriteHourOfSafetyTableNonPreDef
- Added another test case with power outage + natural ventilation. The comparison of with vs without natural ventilation is added to the end of the PR

yujiex committed:
Move annual resilience report data from dataHeatBalFanSys to Zone

- move the arrays like ZoneHeatIndexHourBins  to Zone, so that no separate array
allocation is needed
- moved some int bool fields like ZoneHeatIndex, ZoneNumOcc to Zone as well

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
clang-format

yujiex committed:
Remove unused function

yujiex commented:
Hi @JasonGlazer @Myoldmopar I moved the resilience metrics-related data fields like ZoneHighSETHours from `HeatBalFanSysData` to `ZoneData`. So arrays like these, `state.dataHeatBalFanSys->ZoneHighSETHours`, don't need to be separately allocated. Some fields like ZoneHeatIndex, ZoneNumOcc, etc. are also moved to the `ZoneData`. I didn't move all such fields (the following chunk is just a snapshot of part of such fields) into ZoneData as I'm not sure if that will be too much change and affect other people's code too much.

```
        state.dataHeatBalFanSys->ZTAV.dimension(state.dataGlobal->NumOfZones, 23.0);
        state.dataHeatBalFanSys->ZTAVComf.dimension(state.dataGlobal->NumOfZones, 23.0);
        state.dataHeatBalFanSys->ZT.dimension(state.dataGlobal->NumOfZones, 23.0);
        state.dataHeatBalFanSys->TempTstatAir.dimension(state.dataGlobal->NumOfZones, 23.0);
        state.dataHeatBalFanSys->MAT.dimension(state.dataGlobal->NumOfZones, 23.0);
        state.dataHeatBalFanSys->ZoneTMX.dimension(state.dataGlobal->NumOfZones, 23.0);
        state.dataHeatBalFanSys->ZoneTM2.dimension(state.dataGlobal->NumOfZones, 23.0);
        // Allocate this zone air humidity ratio
        state.dataHeatBalFanSys->ZoneAirHumRatAvg.dimension(state.dataGlobal->NumOfZones, 0.01);
        state.dataHeatBalFanSys->ZoneAirHumRatAvgComf.dimension(state.dataGlobal->NumOfZones, 0.01);
        state.dataHeatBalFanSys->ZoneAirHumRat.dimension(state.dataGlobal->NumOfZones, 0.01);
        state.dataHeatBalFanSys->ZoneAirHumRatOld.dimension(state.dataGlobal->NumOfZones, 0.01);
        state.dataHeatBalFanSys->SumHmAW.dimension(state.dataGlobal->NumOfZones, 0.0);
```

I didn't change the corresponding reporting period data fields, as the size depends on the number of reporting periods from user inputs, and will need to have a separate allocation anyway.

yujiex committed:
clang-format

mjwitte commented on src/EnergyPlus/DataHeatBalance.hh:
```diff
@@ -658,6 +658,42 @@ namespace DataHeatBalance {
         std::vector<ExteriorEnergyUse::ExteriorFuelUsage> otherEquipFuelTypeNums; // List of fuel types used by other equipment in this zone
         std::vector<std::string> otherEquipFuelTypeNames;                         // List of fuel types used by other equipment in this zone
 
+        // People Thermal Resilience Related
+        Real64 ColdStressTempThresh = 15.56;
```
@yujiex It's good to move these into a struct, but the thermal resilience data is somewhat large and only used in limited places, so I don't think we want to lug them around whenever `dataHeatBal->Zone` is accessed. The fields that are used only for thermal resilience should be in their own struct.

And thank you for not touching other arrays. I've started moving some of those into structs in #9394.

Myoldmopar replied:
@mjwitte I had suggested moving some of the standalone Array1D objects into plain scalars on the zone data structure.  Doing this will avoid having separate arrays, each with their own heap allocation, bookkeeping and flags to track what has been allocated and what hasn't...when all that is needed is a flag on the zone data structure.  Is your concern about the list of variables on ZoneData getting too large and hard to track?  Would it be more appealing if all these were collected into a small struct on its own, with a single member on the ZoneData structure?  Like:

```
struct ZoneResilience {
  Real64 someThreshold = 0.0;
  Real64 otherValue = 63.1;
  //...
}

struct ZoneData {
  // ... lots of zone data
  ZoneResilience resilience;
}

// then when you access it in the code, it would be like:
auto &thisZone = ZoneData(ZoneNum):
x = y + thisZone.resilience.someThreshold;
```

That seems like a fairly easy and nice change to make; what do you think?

yujiex replied:
Hi @Myoldmopar, I discussed this with @mjwitte yesterday. The concern is zone data will get too large and potentially unnecessarily drag along a whole bunch of unused data fields when resilience reports are not requested. A separate struct holding only resilience data fields will potentially help with this situation. Or we were also thinking we could even break the resilience table-related struct into subgroups and make them into separate structs if the reliance-related data fields alone are too many to be held in one struct. I was working on investigating the shading feature. I hope to get back to this feature and implement the rearrangement during the weekend.

JasonGlazer commented:
@yujiex  looks like there are some build messages that may need fixing. I tried to update your branch but ran into some merge conflicts. I will leave those to you. I will try to finish up my review soon after you resolve those.

yujiex commented:
> @yujiex looks like there are some build messages that may need fixing. I tried to update your branch but ran into some merge conflicts. I will leave those to you. I will try to finish up my review soon after you resolve those.

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Move resilience table fields into a struct member in ZoneData

yujiex committed:
clang-format

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Fix testfile idf, adding cold/heat temperature threshold field

yujiex committed:
Fix one more idf, adding cold/heat temperature threshold

yujiex committed:
Change position of cold heat stress temperature threshold in idd

yujiex committed:
Use constexpr to specify num of columns in ZoneResilience

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Change report table functions arguments to reference

yujiex committed:
Make ZoneResilience a separate struct

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex committed:
Change WriteResilienceBinsTable templating to number of columns

yujiex committed:
remove redundant function arg columnNum, use the template one

amirroth commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -12971,11 +12971,10 @@ void WriteThermalResilienceTablesRepPeriod(EnergyPlusData &state, int const peri
+                              const std::array<Real64, columnNum> DataHeatBalance::ZoneResilience::*memberPtr,
                               Real64 const unitConvMultiplier)
 {
     std::vector<Real64> columnMax(columnNum, 0);
```
This can be a `std::array<Real64, columnNum>` now, which is the real benefit of making this change.

amirroth commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -13496,13 +13494,11 @@ void WriteThermalResilienceTables(EnergyPlusData &state)
         std::array<Real64, 5> DataHeatBalance::ZoneResilience::*ptrHeatIndexOccupiedHour =
```
I think you have a `constexpr` symbol for 5 somewhere.

amirroth commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -13662,7 +13655,6 @@ void WriteCO2ResilienceTables(EnergyPlusData &state)
     // Using/Aliasing
     if (state.dataGlobal->NumOfZones > 0) {
-        int columnNum = 3;
         std::vector<int> columnHead = {state.dataOutRptPredefined->pdchCO2HourSafe,
```
This can probably be a `std::array` also given that you've change the template variable from `std::array<Real64,columnNum>` to `columnNum`. `columnNum` can be used to parameterize both `std::array<Real64, columnNum>` and `std::array<int, columnNum>`.

yujiex replied:
Yes, make sense, header should have the same size as number of columns, as well as columnMin, columnMax. I'll parameterize all of them with columnNum. I'll also use constexpr in place of the numbers like 5.

yujiex committed:
Use constexpr in array size to parameterize header, minmax

also reduced some indentation by early return

yujiex committed:
clang-format

yujiex committed:
Remove unused function

yujiex committed:
templat columnHead Width in WriteResilienceBinsTable with columnNum

yujiex committed:
remove unused testcase

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

JasonGlazer commented:
@yujiex I did a bunch of review and testing this morning and I think it is almost ready to go. 

- Please fix the clang-format issues for the unit tests
- Please fix the extra warning messages that are being generated "REPORT:TABLE Inconsistant number of columns” which usually means that the data and the columns are not the same.

After that I think it is ready.

yujiex commented:
> @yujiex I did a bunch of review and testing this morning and I think it is almost ready to go.
> 
> * Please fix the clang-format issues for the unit tests
> * Please fix the extra warning messages that are being generated "REPORT:TABLE Inconsistant number of columns” which usually means that the data and the columns are not the same.
> 
> After that I think it is ready.

Thanks, Jason. I'll fix these now.

amirroth commented:
> > @yujiex I did a bunch of review and testing this morning and I think it is almost ready to go.
> > 
> > * Please fix the clang-format issues for the unit tests
> > * Please fix the extra warning messages that are being generated "REPORT:TABLE Inconsistant number of columns” which usually means that the data and the columns are not the same.
> > 
> > After that I think it is ready.
> 
> Thanks, Jason. I'll fix these now.

*inconsistent*.

yujiex committed:
Resolve warning REPORT:TABLE Inconsistent number of columns and typo

yujiex committed:
clang-format

yujiex committed:
Merge remote-tracking branch 'origin/develop' into enhanceResilienceMetrics

yujiex commented:
Hi, @JasonGlazer I've fixed the warning of "REPORT:TABLE Inconsistent number of columns" and also the typo in the warning @amirroth. clang-format is fixed too.

JasonGlazer commented:
Looks like this is ready. Thanks @yujiex !

yujiex commented:
> Looks like this is ready. Thanks @yujiex !

Thanks, @JasonGlazer :)

Myoldmopar commented:
The `integration.AbsorptionChiller_Macro` test is failing on input.  I suspect you need to make a change to the contents in one of the IMF files.  If you need help let me know.

Otherwise this looks clean.  On second thought I'll just try to fix the input while I'm testing out the branch now.  If it takes too much time I'll send this back to you to wrap up.

Myoldmopar committed:
Remove fields from macro file

Myoldmopar commented:
Alright that was easy enough.  You had made changes to the IMF file for the fields, but then when you moved the fields to the end of the IDD object, you forgot to revert those back out.  It passes now, and I am not going to wait on another full round of CI.  I'm merging this, thanks @yujiex for the changes, and @JasonGlazer and @amirroth for reviewing!

yujiex commented:
Thanks, @Myoldmopar. I must have filtered the search relace to idf's when I changed the idd field order. Thanks, @Myoldmopar @JasonGlazer @amirroth @mjwitte @EnergyArchmage for reviewing the branch :)
